### PR TITLE
refactor(moe): refactor moe for extensibility

### PR DESCRIPTION
**Pull request by blankde**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

This PR allow users to implement custom MoE (e.g. megablock moe) more conveniently.

## Modification

1. internlm/moe/base_moe.py: defines the base class of moe layer
2. internlm/moe/sharded_moe.py, internlm/model/moe.py: change class name from MOELayer to GShardMOELayer
3. internlm/utils/model_checkpoint.py: remove suffix for gate key in order to implement the gate strategy more easily

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: internlm/utils/model_checkpoint.py
Comment (ZhiYuanZeng) [inline internlm/utils/model_checkpoint.py:1]: 我有个疑问是我们为什么不能通过all_reduce=False，或者是给param注入一个属性去判断是否是moe参数，例如p.expert=True

Comment (blankde) [reply internlm/utils/model_checkpoint.py:1]: 这种方式也是可以的，从实现上来说区别不大。但是因为这里state_dict()返回的是个字典，用key去判断比较自然一点

**Review by ZhiYuanZeng — APPROVED** [review]

File: internlm/utils/model_checkpoint.py
Comment (ZhiYuanZeng) [reply internlm/utils/model_checkpoint.py:1]: 其实 在有些场景下gate可能会和ep或者expert绑定，也就是说在同一个ep group内的不同进程的gate参数不一样，例如实现一个层次化路由，先在GPU之间分配tokens，然后在GPU内分配tokens，第一层路由的参数是全局共享的，但是第二层参数不是

Comment (blankde) [reply internlm/utils/model_checkpoint.py:1]: 这样的话整个ckpt的保存逻辑就变了，我建议这种情况可以新写一个保存逻辑。

Comment (ZhiYuanZeng) [reply internlm/utils/model_checkpoint.py:1]: 好嘞，那就先这样设置吧

File: internlm/moe/sharded_moe.py
Comment (ZhiYuanZeng) [inline internlm/moe/sharded_moe.py:1]: 可以加上input_split和output_split，all2all是支持不同gpu通信的数据量不一样的
```
class All2All(torch.autograd.Function):
    @staticmethod
    def forward(ctx, xs, input_splits=None, output_splits=None):
        ctx.input_splits = input_splits
        ctx.output_splits = output_splits

        ys = (
            torch.empty_like(xs)
            if output_splits is None
            else xs.new_empty(size=[sum(output_splits)] + list(xs.size()[1:]))
        )
        torch.distributed.all_to_all_single(
            ys, xs, output_split_sizes=output_splits, input_split_sizes=input_splits
        )
        return ys

    @staticmethod
    def backward(ctx, grad_output):
        result = (
            torch.empty_like(grad_output)
            if ctx.input_splits is None
            else grad_output.new_empty(
                size=[sum(ctx.input_splits)] + list(grad_output.size()[1:])
            )
        )
        torch.distributed.all_to_all_single(
            result,
            grad_output,
            output_split_sizes=ctx.input_splits,
            input_split_sizes=ctx.output_splits,
        )
        return result, None, None
```

**Review by ZhiYuanZeng — COMMENTED** [review]
MoE类的定义能不能更通用一点，因为top1, top2, capacity factor都是Gshard的参数，其它的路由可能没有这些参数

**Comment by blankde** [comment]
> MoE类的定义能不能更通用一点，因为top1, top2, capacity factor都是Gshard的参数，其它的路由可能没有这些参数

@ZhiYuanZeng 致远，这个commit（https://github.com/InternLM/InternLM/pull/567/commits/fe0c342f9dfd99fe1ac7e828e68a45f6db68356e）把top1, top2, capacity factor的参数都放在了GshardMoE里面，目前MoE类的定义只包含了必要的信息，你看下是否可以

**Comment by ZhiYuanZeng** [comment]
> > MoE类的定义能不能更通用一点，因为top1, top2, capacity factor都是Gshard的参数，其它的路由可能没有这些参数
> 
> @ZhiYuanZeng 致远，这个commit [fe0c342](https://github.com/InternLM/InternLM/commit/fe0c342f9dfd99fe1ac7e828e68a45f6db68356e) 把top1, top2, capacity factor的参数都放在了GshardMoE里面，目前MoE类的定义只包含了必要的信息，你看下是否可以

需要调整 `build_model_with_moe_cfg`适应这个变化吗

**Comment by blankde** [comment]
> > > MoE类的定义能不能更通用一点，因为top1, top2, capacity factor都是Gshard的参数，其它的路由可能没有这些参数
> > 
> > 
> > @ZhiYuanZeng 致远，这个commit [fe0c342](https://github.com/InternLM/InternLM/commit/fe0c342f9dfd99fe1ac7e828e68a45f6db68356e) 把top1, top2, capacity factor的参数都放在了GshardMoE里面，目前MoE类的定义只包含了必要的信息，你看下是否可以
> 
> 需要调整 `build_model_with_moe_cfg`适应这个变化吗

兼容之前的modeling，不用调整build_model_with_moe_cfg

**Comment by blankde** [comment]
@ZhiYuanZeng 重构了moe类， https://github.com/InternLM/InternLM/pull/567/commits/7cec7e985f392d9be6a839c455f3c39e25a24a90

**Review by ZhiYuanZeng — APPROVED** [review]

**Comment by blankde** [comment]
@gaoyang07 @ZhiYuanZeng 我在最新的这个commit https://github.com/InternLM/InternLM/pull/567/commits/8acf8455fe6f50565d6017536c8b52a555d64c18 修改了一下获取moe具体实现的判断，复用了之前model regester的实现，佬们看下是否可以？

**Pull request closed without merging**